cachi2/core/package_managers/rpm/redhat.py

```diff
@@ -0,0 +1,168 @@
+import asyncio
+import hashlib
+import logging
+import uuid
+from collections import UserDict
+from os import PathLike
+from pathlib import Path
+from typing import Any, Optional, Union
+
+from pydantic import BaseModel, PositiveInt, ValidationError
+
+from cachi2.core.config import get_config
+from cachi2.core.package_managers.general import async_download_files
+
+log = logging.getLogger(__name__)
+
+
+class Package(BaseModel):
+    """Package item; represents RPM or SRPM file. Used for parsing the lockfile."""
+
+    repoid: Optional[str] = None
+    url: str
+    checksum: Optional[str] = None
+    size: Optional[int] = None
+
+
+class Arch(BaseModel):
+    """Architecture structure. Used for parsing the lockfile."""
+
+    arch: str
+    packages: list[Package] = []
+    source: list[Package] = []
+
+
+class Root(BaseModel):
+    """Top of the structure of the lockfile. Used for parsing the lockfile."""
+
+    lockfileVersion: PositiveInt
+    lockfileVendor: str
+    arches: list[Arch]
+
+
+class Header(BaseModel):
+    """
+    The class represents header of the lockfile. Used for matching the format.
+
+    Contains just a minimal set of fields to identify the format.
+    """
+
+    lockfileVersion: PositiveInt
+    lockfileVendor: str
+
+
+class RedhatRpmsLock(UserDict):
+    """
+    The class implements basic operations with specific lockfile format.
+
+    Input data comes from raw yaml stored as a dictionary.
+    """
+
+    # parsed structure of the lockfile
+    _lockfile: Root
+    # metadata of file records found in the lockfile (for verification and SBOM generation)
+    _files_metadata: dict[Path, Any] = {}
```

Comment:
This attribute seems to be just duplication of data we already hold via (currently via the `_lockfile` attribute, which will go away).

Reply:
Not sure what the comment is related to:

* _files_metadata
We have the data in the loaded/parsed structure from lockfile. But `_files_metadata` (now `metadata`) is organized differently (the main key is the package path). It makes iterating through items much easier (downloading, verification after download, SBOM generation). Otherwise, for example, I would have to construct paths multiple times (including testing whether the path contains internal repoid). I prefer to keep metadata.

* _lockfile
Yes, it will disappear.